doc/mgmt/config_replace_rollback_HLD.md

```diff
@@ -0,0 +1,418 @@
+- Config replace operation should be performed with minimum disruption to the device, i.e. no container restart or system reboots expected
+- Minimize the traffic impact only on the path of config change without affecting any unmodified config
+- Default configuration, e.g. copp_config.json, not to be saved to a serialized DB file should be preserved in the CONFIG_DB across config replacement operation.
+- Media format of intended config files can be a raw json, tar compresses, or special format defined in configure session proposal.
+
+**Configure session CLI**
+- A new CLI (replace) command to perform configuration replace from configure session mode.
+- A "timer" option added to commit command for user to apply the change to CONFIG_DB on trial basis, if a timer gets expired, rollback to the saved backup file.
+	+ While CONFIG_DB is in trial stage, config changes are not allowed but read operation still allowed.
+- A "confirm" option added to commit command for user to confirm new configuration changes, i.e. move trial to permanent stage.
+- When "Kill" is executed
+	+ timer will be canceled and running-configuration reverted back to previous configuration state.
+
+**Non configure session CLI**
+- A "replace" option added to existing CLI "copy source-url running-configuration" command to perform configuration replacement on exec mode.
+
+**REST/gNMI/gNOI** 
+- A new REST POST operation to perform config replacement.
+- A new gNMI/gNOI operation to perform config replacement.
+
+
+## 1.3 Design Overview
+### 1.3.1 Basic Approach
+The config replace design will leverage the following configure session functionality as well as extend it as specified.
+
+- When a configure session is opened, CONFIG_DB is locked and an unique session token is assigned to the requests from config session to differentiate from non-session based requests. DB writes will be rejected if requests are initiated from REST, gNMI/gNOI including non-configure session CLI.
+- A session token is used for creating, viewing checkpoints optionally with label. Non-session based operations will not be using such.
+- Commit command for cached candidate datastore is non-modeled CLI-only RPCs. Extension required for configure replace use, e.g. commit timer, commit confirm
+- All the changes within a session will be cached to candidate datastore, not to CONFIG_DB until a commit (with or without timer) is issued. Config replace from non configure session mode is to apply changes directly to CONFIG_DB.
+
+
+When a configure replace resquest is sent from non-configure session or REST, gNMI/gNOI:
+- Check if CONFIG_DB is locked. If it is locked, the command is rejected to avoid conflicting configuration state.
+
+When a configure replace request is sent within a configure session:
+- When 'replace' command is issued, check if the candidate store is dirtified. If it is dirtified, the command is rejected to avoid conflicting configuration state.
```

Comment:
What is meant by dirtified?

Reply:
DB is not clean, indicating that has changes. It is a term used while in discussion for config management.